File: 347-top-k-frequent-elements/top-k-frequent-elements.cpp
class Solution {
public:
    vector<int> topKFrequent(vector<int>& nums, int k) {
        unordered_map<int, int> count; // Count the frequency of each element
        for (auto &i : nums) {
            count[i]++;
        }

        // Create buckets where index represents frequency
        vector<vector<int>> bucket(nums.size() + 1);
        for (const auto &ele : count) {
            bucket[ele.second].push_back(ele.first); // Group numbers by frequency
        }

        vector<int> res;
        // Iterate through the buckets from the end (highest frequency)
        for (int i = bucket.size() - 1; i >= 0 && res.size() < k; --i) {
            for (const auto &num : bucket[i]) {
                res.push_back(num); // Collect top k frequent elements
                if (res.size() == k) break;
            }
        }

        return res;
    }
};

File: 347-top-k-frequent-elements/top-k-frequent-elements.py
import heapq
class Solution(object):
    def topKFrequent(self, nums, k):
        """
        :type nums: List[int]
        :type k: int
        :rtype: List[int]
        """
        # count freq
        freq = {}
        for num in nums:
            if num in freq:
                freq[num] += 1
            else:
                freq[num] = 1
        # freq = {1:3, 2:2, 3:1}
        # how to use heap? 
        # using min-heap of size k
        heap = []
        for key in freq:
            val = freq[key]
            heapq.heappush(heap, (val, key))
            if len(heap) > k:
                heapq.heappop(heap)

        # from here, heap only contains top k elements
        # res = []
        # for (key, val) in heap:
        #     res.append(val)

        res = [val for (key, val) in heap]
        return res


        